File: Civitas - Ruby/lib/sorpresa_casilla.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.

module Civitas
  class Sorpresa_casilla < Sorpresa
    public_class_method :new
    def initialize(tablero, valor, texto)
      super texto
      @tablero = tablero
      @valor = valor
    end
    
    #Aplica la carta de ir a la casilla especificada por la carta
    def aplicar_a_jugador(actual, todos)
      if(jugador_correcto(actual, todos))
        informe(actual, todos)
        casilla_act = todos[actual].num_casilla_actual
        tirada = @tablero.calcular_tirada(casilla_act, @valor)
        nueva_pos = @tablero.nueva_posicion(casilla_act, tirada)
        todos[actual].mover_a_casilla(nueva_pos)
        @tablero.get_casilla(nueva_pos).recibe_jugador(actual, todos)
      end
    end
    
    def to_s
      return super + "\nValor: " + @valor
    end
    
  end
end

File: Civitas - Java/src/civitas/SorpresaCasilla.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;

/**
 *
 * @author juanc
 */
public class SorpresaCasilla extends Sorpresa{
    
    //Para ir a otra casilla
    SorpresaCasilla (Tablero tablero, int valor, String texto){
        super(valor,texto,tablero);
    }
        
    //Aplica la carta de ir a la casilla especificada por la carta
    @Override
    void aplicarAJugador(int actual, ArrayList<Jugador> todos){
        
        if(jugadorCorrecto(actual, todos) ){
            informe(actual, todos);
            int cas_actual = todos.get(actual).getNumCasillaActual();
            int tirada = tablero.calcularTirada(cas_actual, valor);
            int nueva_pos = tablero.nuevaPosicion(cas_actual, tirada);
            
            todos.get(actual).moverACasilla(nueva_pos);
            tablero.getCasilla(nueva_pos).recibeJugador(actual, todos);
        }
        
    }
}
